lib/nm/index.ts

```diff
@@ -0,0 +1,339 @@
+
+	requestScan = (params) => this.callMethod(this.devices.wifi.path)('org.freedesktop.NetworkManager.Device.Wireless')(['RequestScan', 'a{sv}'])([params]);
+
+	getActiveConnections = () => this.callMethod()('org.freedesktop.DBus.Properties')(['Get', 'ss'])(['org.freedesktop.NetworkManager', 'ActiveConnections']);
+
+	getConnectionSettings = async (path) => {
+		try {
+			const value = await this.callMethod(path)(['Settings', 'Connection'])(['GetSettings', ''])();
+			return { path, settings: value };
+		} catch (err) {
+			throw formatError(500, `Could not getConnectionSettings on ${path}`, err);
+		}
+	};
+
+	connectNetwork = async (network) => {
+		const connectionParam = [
+			['connection', [
+				['id', ['s', network.ssid]],
+				['type', ['s', '802-11-wireless']],
+			]],
+			['802-11-wireless', [
+				['ssid', ['ay', [stringToArrayOfBytes(network.ssid)]]],
+				['mode', ['s', 'infrastructure']],
+			]],
+			['802-11-wireless-security', [
+				['auth-alg', ['s', 'open']],
+				['key-mgmt', ['s', 'wpa-psk']],
+				['psk', ['s', network.passphrase]],
+			]],
+			['ipv4', [
+				['method', ['s', 'auto']],
+			]],
+			['ipv6', [
+				['method', ['s', 'auto']],
+			]],
+		];
+
+		const ap: AccessPoint|any = _.find(this.accessPoints, (props) => {
+			return props.Ssid.toString() === network.ssid;
+		});
+		if (_.isUndefined(ap)) {
+			return Bluebird.reject(formatError(404, `Could not find neaby AccessPoints with SSID: ${network.ssid}`));
+		}
+
+		try {
+			// Check if network's already registered on the device
+			const results = await this.listConnections();
+			const wifiConnection = findConnection(results, network);
+			if (!_.isUndefined(wifiConnection)) {
+				await this.activateConnection([wifiConnection.path, this.devices.wifi.path, '/']);
+				return;
+			}
+
+			const networkSettings = await this.addConnection(connectionParam);
+			await this.activateConnection([networkSettings, this.devices.wifi.path, '/']);
+			return;
+		} catch (err) {
+			throw formatError(500, 'Could not connectNetwork', err);
+		}
+	}
+
+	listConnections = async () => {
+		try {
+			const connections = await this.callMethod(['Settings'])(['Settings'])(['ListConnections', ''])();
+			const getConnectionsSettings = () => _.map(connections, this.getConnectionSettings);
+			const results = await Bluebird.all(getConnectionsSettings());
+			return results;
+		} catch (err) {
+			throw formatError(500, `Could not listConnections`, err);
+		}
+	}
+
+	forgetNetwork = async (network) => {
+		try {
+			const results = await this.listConnections();
+			const wifiConnection = findConnection(results, network);
+			await this.deleteConnection(wifiConnection.path);
+			return Bluebird.resolve();
+		} catch (err) {
+			throw formatError(500, `Could not forgetNetwork`, err);
+		}
+	}
+
+	listNearbyNetworks = async () => {
+		try {
+			const requestScanParams = [
+				['ssids', ['aay', [stringToArrayOfBytes('1')]]],
+			];
+			await this.requestScan(requestScanParams);
```

Comment:
After the request scan is completed here will the actual scan be completed? Asking because this may just trigger the scan, but he results can be available after a few seconds and the access point list below could be retrieved before that.

Reply:
Good point: usually I get the complete list after a second scan, which cannot take place before ~5 seconds have passed.
A solution would be to listen to the `AccessPointAdded`/`-Removed`" events, but that would also mean delegating the `RequestScan` logic to an `interval`led function and keeping track of the `AccessPoint`s separately, which is not bad but I feel more comfortable with a "give me the list -> ask NM list -> return list" logic.
Do you have any suggestion @majorz on how to handle this more properly?

Reply:
Side note: listening for those two events would also mean having something like a `toggleScan` function to trigger the scans only when the user actually wants to retrieve the list (e.g. when he is looking at the wifi settings page with the list to be updated by `listConnections`)

Reply:
I will dig a bit into the NM source code to see what provided options we have. wpa_supplicant has a ScanDone signal, which I can use to track NM its usage.

Reply:
@majorz I don't think we have the option to use wpa_supplicant1, and as far as the documentation goes, there's no similar event in NetworkManager...
What I can alternatively do is a combination of the two options I provided: I can still keep trying to do a `RequestScan`, and if it fails, just return the `AccessPoints` property, so the user can keep doing that as long as he wants.
I kinda like this, what do you think?

Reply:
What I mean is check for these two errors:
`Scanning not allowed while already scanning` and `Scanning not allowed immediately following previous scan`
If the error equals one of those, then I think I can safely return just the AP list already provided by NM

Reply:
UPDATE: Just tried it, looks pretty clean and no strange workarounds to circumvent those two errors (might as well consider them more warnings than errors, since they don't cause anything weird/bad)

Reply:
I meant I will use it to trace how NM handles this internally in the source code. So the `PropertiesChanged` signal from `org.freedesktop.DBus.Properties` can be used on the `Device.Wireless` interface for determining when the scan has completed. This is just for information.

I suggest that we do not use RequestScan inside the listNearbyNetworks function, but just have it available separately. It can be called once on initialization. Then the listNearbyNetworks function can be called from the UI with some interval, so that the list with networks is updated regularly in the UI. What do you think?

Reply:
@majorz This sounds good as long as there is no active wifi connection, but as soon as a network is activated, the automatic scans happen only once every 3 minutes and the returned list contains only the currently connected network, meaning that we *need* to do some manual `RequestScan` one way or another to update a `Device.Wireless`'s `AccessPoints` property with nearby networks

Reply:
For reference, the fix would be this:
```javascript
if (_.isArray(err) && (err[0] === 'Scanning not allowed while already scanning' || err[0] === 'Scanning not allowed immediately following previous scan')) {
  return this.getAccessPoints(this.devices.wifi.path);
}
```

Reply:
Right, indeed when there is an active wifi connection the scan will be disabled. I am not sure it will even happen on 3 minutes interval. I think it makes sense to be called once in the while from the UI.

The scan completes in less than two seconds from what I have seen.

Reply:
Yeah and I already added a check on those two errors, as well as another one I discovered while using the module more intensely (cannot scan right after enabling, or something similar)
Sorry for the late answer, I didn't see your comment on this @jhermsmeier !